sql: panic: column 2 not in input

Setup:

```
CREATE TABLE table9 (col0 OID[] NULL, col1 TIMESTAMP, col2 TIME, col3 UUID NOT NULL, col4 NAME NOT NULL, col5 DECIMAL NULL, col6 DECIMAL NOT NULL, col7 DECIMAL NOT NULL, col8 DATE NULL, col9 OID[], col10 INTERVAL NOT NULL, col11 VARBIT, col12 OID[] NULL, FAMILY fam0 (col10, col11), FAMILY fam1 (col4), FAMILY fam2 (col9), FAMILY fam3 (col0), FAMILY fam4 (col3), FAMILY fam5 (col1, col5), FAMILY fam6 (col8), FAMILY fam7 (col6), FAMILY fam8 (col12), FAMILY fam9 (col7, col2), PRIMARY KEY (col11 ASC, col10 ASC), UNIQUE (col4 ASC), INDEX (col5 DESC, col1 ASC, col10, col11, col6, col4, col3 ASC, col8 DESC), UNIQUE (col2 ASC), INDEX (col7 ASC, col4, col5 ASC, col10 DESC), INDEX (col6 ASC, col3), UNIQUE (col7, col6));
CREATE TABLE table6 (col0 TIMESTAMP, col1 NAME NULL, col2 INET, col3 INET, col4 TIMESTAMP, col5 NAME, col6 TIME, col7 DECIMAL, col8 INT8, col9 TIMESTAMP NULL, col10 STRING COLLATE da, col11 TIMESTAMP NOT NULL, col12 NAME, col13 BOOL NOT NULL, col14 STRING COLLATE da NULL, FAMILY fam0 (col4, col2, col12, col5, col3, col9, col0, col1, col14, col7, col10, col6, col11, col8, col13), PRIMARY KEY (col5, col7, col9, col14 ASC, col0 ASC, col1 DESC, col4, col6 DESC, col3, col12 DESC, col11, col2), UNIQUE (col11 ASC, col7 DESC, col14 ASC, col2 ASC, col4 DESC, col12 DESC, col13 ASC), INDEX (col10 DESC, col0 DESC, col4 DESC, col2 ASC, col8 DESC, col11 ASC, col5 DESC, col6 DESC, col1 ASC, col13, col12 ASC, col9, col7), INDEX (col1, col6, col2 ASC, col14, col5, col10 ASC, col12 DESC, col11), INDEX (col0 DESC), UNIQUE (col5 ASC, col1, col9, col2 ASC, col0 DESC, col8 ASC, col4 DESC, col3 ASC, col13 DESC, col10 ASC, col6 ASC, col12, col7 ASC, col11 DESC, col14 DESC), INDEX (col10 ASC, col13 DESC, col5, col7 ASC, col3, col1, col4, col8 ASC, col11 ASC), UNIQUE (col9, col11, col8 DESC, col13 DESC, col14, col4 DESC, col7 DESC, col6 DESC, col5 ASC), UNIQUE (col2, col8 DESC, col6 DESC, col5 ASC, col14 DESC, col13 DESC, col10 ASC, col0 ASC, col3 ASC, col9 ASC, col7 ASC));
```

Query:

```
SELECT
	*
FROM
	(
		VALUES
			(2192725121:::OID),
			(
				(
					SELECT
						col_1082927 AS col_1082930
					FROM
						[
							INSERT
							INTO
								defaultdb.public.table9
									AS tab_362035
									(
										col3,
										col1,
										col4,
										col7,
										col8,
										col10,
										col11,
										col6
									)
							SELECT
								'7f69c091-17c8-4578-b197-20edbc5c5427':::UUID
									AS col_1082917,
								tab_362036.col11
									AS col_1082918,
								tab_362036.col1
									AS col_1082919,
								tab_362036.col7
									AS col_1082920,
								'1995-01-06':::DATE
									AS col_1082921,
								'28 years 8 mons 661 days 18:05:36.156337':::INTERVAL
									AS col_1082922,
								B'001000100010010011100101110100'
									AS col_1082923,
								tab_362036.col7
									AS col_1082924
							FROM
								defaultdb.public.table6
									AS tab_362036
							RETURNING
								col2 AS col_1082925,
								col1 AS col_1082926,
								906124112:::OID
									AS col_1082927,
								NULL AS col_1082928,
								CASE
								WHEN true
								THEN '1970-01-12 10:21:04.000328+00:00':::TIMESTAMPTZ
								ELSE '1970-01-02 15:12:27.000655+00:00':::TIMESTAMPTZ
								END
									AS col_1082929
						]
					ORDER BY
						col_1082926 DESC, col_1082925 ASC
					LIMIT
						1:::INT8
				)
			),
			(176065225:::OID),
			(2630168490:::OID),
			(3840633645:::OID),
			(
				CASE
				WHEN (
					'd9ee:3044:68a8:ee29:2c2c:b239:d3eb:4291/19':::INET
					>> '46.169.10.60/19':::INET
				)
				THEN 3185974125:::OID
				ELSE 1471795872:::OID
				END
			)
	)
		AS tab_362037 (col_1082931);
```

Trace:

```
panic: column 2 not in input [recovered]
        panic: panic while executing 1 statements: SELECT * FROM (VALUES (_:::OID), (__more5__)) AS _ (_); caused by column 2 not in input

goroutine 173 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0008d4000, 0x3a32f20, 0xc000979200, 0x2d53320, 0xc00127e1b0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:715 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc0008d4000, 0x3a32f20, 0xc000979200)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:427 +0x61
panic(0x2d53320, 0xc00127e1b0)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*execPlan).getColumnOrdinal(0xc001aabca8, 0x2, 0x1)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:90 +0x193
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*execPlan).sqlOrdering(0xc001aabca8, 0xc0003904b4, 0x1, 0x1, 0x0, 0xc0012807e0, 0xc001795180)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:117 +0x91
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).ensureColumns(0xc0009452c0, 0x3164e60, 0xc0011ef680, 0x0, 0x0, 0xc0012807e0, 0xc001795180, 0xd, 0xd, 0x0, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:1635 +0x19c
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildInsert(0xc0009452c0, 0xc000663070, 0x0, 0x0, 0x0, 0x0, 0x0, 0x17, 0x6b8a4c)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:1300 +0x24f
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc0009452c0, 0x3a87b20, 0xc000663070, 0x0, 0x0, 0x0, 0x0, 0x0, 0x6b8a4c, 0xc00003c000)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:210 +0xe9c
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildLimitOffset(0xc0009452c0, 0x3a87e40, 0xc0011d1570, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc001aac038, 0x1b566da)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:1003 +0xba
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc0009452c0, 0x3a87e40, 0xc0011d1570, 0x0, 0x0, 0x0, 0x0, 0x0, 0x2f90073, 0x304b420)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:177 +0x6bc
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildProject(0xc0009452c0, 0xc0011d1970, 0x0, 0x0, 0x0, 0x0, 0x0, 0x6b8c16, 0x6b8a4c)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:439 +0x8a
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc0009452c0, 0x3a88160, 0xc0011d1970, 0x0, 0x0, 0x0, 0x0, 0x0, 0x3a37ce0, 0xc00170da40)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:168 +0x64d
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).build(0xc0009452c0, 0x3a42120, 0xc0011d1970, 0x3a42120, 0xc0011d1970, 0x3a58200, 0xc00170da40)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:68 +0x89
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildSubquery(0xc0009452c0, 0xc00108a100, 0x3a60120, 0xc001717e00, 0xc00170da40, 0x0, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:547 +0x13b
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc0009452c0, 0xc00108a100, 0x3a60120, 0xc001717e00, 0x3a58200, 0xc00170da40, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:94 +0xa6
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildValues(0xc0009452c0, 0xc0011d1b70, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x6b8a4c)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:268 +0x1fd
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc0009452c0, 0x3a88c00, 0xc0011d1b70, 0x0, 0x0, 0x0, 0x0, 0x0, 0x30e52c0, 0x6bc200)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:156 +0xd85
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).build(0xc0009452c0, 0x3a429e0, 0xc0011d1b70, 0x671, 0xc001aac800, 0x6bc2d8, 0x60)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:68 +0x89
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).Build(0xc0009452c0, 0xc0009452c0, 0xc001933740, 0xc00132c2c0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:56 +0x41
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc0008d4350, 0x3a32fe0, 0xc001933740, 0x1, 0xc000000002, 0x3270c7a, 0x19)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:160 +0x196
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).makeExecPlan(0xc0008d4000, 0x3a32fe0, 0xc001933740, 0xc0008d4350, 0x0, 0x855cf89)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:975 +0x154
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0008d4000, 0x3a32fe0, 0xc001933740, 0xc0008d4350, 0x7f4cdb886bd8, 0xc00103a160, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:864 +0x125
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0008d4000, 0x3a32fe0, 0xc001933740, 0x3a38aa0, 0xc0018e2a00, 0xc0014f2000, 0x671, 0x0, 0x0, 0x0, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:457 +0xefc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0008d4000, 0x3a32fe0, 0xc001933740, 0x3a38aa0, 0xc0018e2a00, 0xc0014f2000, 0x671, 0x0, 0x0, 0x0, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:103 +0x638
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0008d4000, 0x3a32f20, 0xc000979200, 0xc000bfe7b8, 0x5400, 0x15000, 0xc000bfe850, 0xc000328160, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1178 +0x2110
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000713ba0, 0x3a32f20, 0xc000979200, 0xc0008d4000, 0x5400, 0x15000, 0xc000bfe850, 0xc000328160, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:429 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc000713ba0, 0xc00079a000, 0xc0003aa0a0, 0xc000328160, 0x3a32f20, 0xc000979200, 0xc0008d4000, 0x5400, 0x15000, 0xc000bfe850, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:332 +0xcf
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:315 +0x1121
```

Found with sqlsmith.

--- a/pkg/sql/opt/exec/execbuilder/relational_builder.go
+++ b/pkg/sql/opt/exec/execbuilder/relational_builder.go
@@ -1297,7 +1297,7 @@ func (b *Builder) buildInsert(ins *memo.InsertExpr) (execPlan, error) {
 	colList := make(opt.ColList, 0, len(ins.InsertCols)+len(ins.CheckCols))
 	colList = appendColsWhenPresent(colList, ins.InsertCols)
 	colList = appendColsWhenPresent(colList, ins.CheckCols)
-	input, err = b.ensureColumns(input, colList, nil, ins.ProvidedPhysical().Ordering)
+	input, err = b.ensureColumns(input, colList, nil, ins.Input.ProvidedPhysical().Ordering)
 	if err != nil {
 		return execPlan{}, err
 	}
@@ -1348,7 +1348,7 @@ func (b *Builder) buildUpdate(upd *memo.UpdateExpr) (execPlan, error) {
 	colList = appendColsWhenPresent(colList, upd.FetchCols)
 	colList = appendColsWhenPresent(colList, upd.UpdateCols)
 	colList = appendColsWhenPresent(colList, upd.CheckCols)
-	input, err = b.ensureColumns(input, colList, nil, upd.ProvidedPhysical().Ordering)
+	input, err = b.ensureColumns(input, colList, nil, upd.Input.ProvidedPhysical().Ordering)
 	if err != nil {
 		return execPlan{}, err
 	}
@@ -1414,7 +1414,7 @@ func (b *Builder) buildUpsert(ups *memo.UpsertExpr) (execPlan, error) {
 		colList = append(colList, ups.CanaryCol)
 	}
 	colList = appendColsWhenPresent(colList, ups.CheckCols)
-	input, err = b.ensureColumns(input, colList, nil, ups.ProvidedPhysical().Ordering)
+	input, err = b.ensureColumns(input, colList, nil, ups.Input.ProvidedPhysical().Ordering)
 	if err != nil {
 		return execPlan{}, err
 	}
@@ -1473,7 +1473,7 @@ func (b *Builder) buildDelete(del *memo.DeleteExpr) (execPlan, error) {
 	// Upgrade execution engine to not require this.
 	colList := make(opt.ColList, 0, len(del.FetchCols))
 	colList = appendColsWhenPresent(colList, del.FetchCols)
-	input, err = b.ensureColumns(input, colList, nil, del.ProvidedPhysical().Ordering)
+	input, err = b.ensureColumns(input, colList, nil, del.Input.ProvidedPhysical().Ordering)
 	if err != nil {
 		return execPlan{}, err
 	}

--- a/pkg/sql/opt/exec/execbuilder/testdata/insert
+++ b/pkg/sql/opt/exec/execbuilder/testdata/insert
@@ -467,3 +467,30 @@ count             ·              ·                      ()                    
 
 statement ok
 ROLLBACK
+
+# Regression test for #35564: make sure we use the Insert's input required
+# ordering for the internal projection.
+
+statement ok
+CREATE TABLE abc (a INT, b INT, c INT, INDEX(c) STORING(a,b))
+
+statement ok
+CREATE TABLE xyz (x INT, y INT, z INT)
+
+query TTTTT
+EXPLAIN (VERBOSE) SELECT * FROM [INSERT INTO xyz SELECT a, b, c FROM abc RETURNING z] ORDER BY z
+----
+render                    ·         ·                    (z)                       +z
+ │                        render 0  z                    ·                         ·
+ └── run                  ·         ·                    (x, y, z, rowid[hidden])  ·
+      └── insert          ·         ·                    (x, y, z, rowid[hidden])  ·
+           │              into      xyz(x, y, z, rowid)  ·                         ·
+           │              strategy  inserter             ·                         ·
+           └── render     ·         ·                    (a, b, c, column9)        +c
+                │         render 0  a                    ·                         ·
+                │         render 1  b                    ·                         ·
+                │         render 2  c                    ·                         ·
+                │         render 3  unique_rowid()       ·                         ·
+                └── scan  ·         ·                    (a, b, c)                 +c
+·                         table     abc@abc_c_idx        ·                         ·
+·                         spans     ALL                  ·                         ·

--- a/pkg/sql/opt/exec/execbuilder/testdata/update
+++ b/pkg/sql/opt/exec/execbuilder/testdata/update
@@ -296,3 +296,29 @@ Del /Table/57/1/1/1/1
 Del /Table/57/1/1/2/1
 fast path completed
 rows affected: 1
+
+# Regression test for #35564: make sure we use the Update's input required
+# ordering for the internal projection.
+
+statement ok
+CREATE TABLE abc (a INT, b INT, c INT, INDEX(c) STORING(a,b))
+
+query TTTTT
+EXPLAIN (VERBOSE) SELECT * FROM [ UPDATE abc SET a=c RETURNING a ] ORDER BY a
+----
+render                    ·         ·              (a)                       +a
+ │                        render 0  a              ·                         ·
+ └── run                  ·         ·              (a, b, c, rowid[hidden])  ·
+      └── update          ·         ·              (a, b, c, rowid[hidden])  ·
+           │              table     abc            ·                         ·
+           │              set       a              ·                         ·
+           │              strategy  updater        ·                         ·
+           └── render     ·         ·              (a, b, c, rowid, c)       +c
+                │         render 0  a              ·                         ·
+                │         render 1  b              ·                         ·
+                │         render 2  c              ·                         ·
+                │         render 3  rowid          ·                         ·
+                │         render 4  c              ·                         ·
+                └── scan  ·         ·              (a, b, c, rowid[hidden])  +c
+·                         table     abc@abc_c_idx  ·                         ·
+·                         spans     ALL            ·                         ·

--- a/pkg/sql/opt/exec/execbuilder/testdata/upsert
+++ b/pkg/sql/opt/exec/execbuilder/testdata/upsert
@@ -307,3 +307,38 @@ INSERT INTO t5 VALUES (1, 10, 9) ON CONFLICT (k) DO NOTHING
 
 statement ok
 INSERT INTO t5 VALUES (1, 10, 20) ON CONFLICT (k) DO NOTHING
+
+# Regression test for #35564: make sure we use the Upsert's input required
+# ordering for the internal projection.
+
+statement ok
+CREATE TABLE abc (a INT, b INT, c INT, INDEX(c) STORING(a,b))
+
+statement ok
+CREATE TABLE xyz (x INT, y INT, z INT)
+
+query TTTTT
+EXPLAIN (VERBOSE) SELECT * FROM [UPSERT INTO xyz SELECT a, b, c FROM abc RETURNING z] ORDER BY z
+----
+render                         ·         ·                    (z)                          +z
+ │                             render 0  z                    ·                            ·
+ └── run                       ·         ·                    (x, y, z, rowid[hidden])     ·
+      └── upsert               ·         ·                    (x, y, z, rowid[hidden])     ·
+           │                   into      xyz(x, y, z, rowid)  ·                            ·
+           │                   strategy  opt upserter         ·                            ·
+           └── render          ·         ·                    (a, b, c, column9, a, b, c)  +c
+                │              render 0  a                    ·                            ·
+                │              render 1  b                    ·                            ·
+                │              render 2  c                    ·                            ·
+                │              render 3  column9              ·                            ·
+                │              render 4  a                    ·                            ·
+                │              render 5  b                    ·                            ·
+                │              render 6  c                    ·                            ·
+                └── render     ·         ·                    (column9, a, b, c)           +c
+                     │         render 0  unique_rowid()       ·                            ·
+                     │         render 1  a                    ·                            ·
+                     │         render 2  b                    ·                            ·
+                     │         render 3  c                    ·                            ·
+                     └── scan  ·         ·                    (a, b, c)                    +c
+·                              table     abc@abc_c_idx        ·                            ·
+·                              spans     ALL                  ·                            ·

